src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -24,71 +28,93 @@ public static int CacheSize
+        ///  Find cache based on options+pattern+culture and optionally add new cache if not found
         /// </summary>
-        private static CachedCodeEntry LookupCachedAndUpdate(CachedCodeEntryKey key)
+        private CachedCodeEntry GetCachedCode(CachedCodeEntryKey key, bool isToAdd)
         {
+            if (s_cacheSize == 0)
+                return null;
             lock (s_livecode)
             {
-                for (LinkedListNode<CachedCodeEntry> current = s_livecode.First; current != null; current = current.Next)
+                // first look for it in the cache and move it to the head
+                var entry = LookupCachedAndPromote(key);
+                // it wasn't in the cache, so we'll add a new one
+                if (entry == null && isToAdd && s_cacheSize != 0)  // check cache size again in case it changed
                 {
-                    if (current.Value._key == key)
+                    entry = new CachedCodeEntry(key, capnames, capslist, _code, caps, capsize, _runnerref, _replref);
+                    s_livecode.Add(key, entry);
+                    // put in linked list:
+                    if (s_livecode_first != null)
+                    {
+                        s_livecode_first._next = entry;
+                        entry._previous = s_livecode_first;
+                    }
+                    s_livecode_first = entry;
+                    if (s_livecode_last == null)
+                    {
+                        s_livecode_last = entry;
+                    }
+                    else
                     {
-                        // If we find an entry in the cache, move it to the head at the same time.
-                        s_livecode.Remove(current);
-                        s_livecode.AddFirst(current);
-                        return current.Value;
+                        if (s_livecode.Count > s_cacheSize)
+                        {
+                            var last = s_livecode_last;
+                            s_livecode.Remove(last._key);
+
+                            last._next._previous = null;
+                            s_livecode_last = last._next;
+                            last._next = null;
+                        }
                     }
                 }
+                return entry;
             }
-
-            return null;
         }
 
-        /// <summary>
-        /// Add current code to the cache
-        /// </summary>
-        private CachedCodeEntry CacheCode(CachedCodeEntryKey key)
+        // Always called within s_livecode lock
+        private static CachedCodeEntry LookupCachedAndPromote(CachedCodeEntryKey key)
         {
-            CachedCodeEntry newcached = null;
-
-            lock (s_livecode)
+            if (s_livecode_first?._key == key) // most used regex should be at the top already
+                return s_livecode_first;
+            s_livecode.TryGetValue(key, out var entry);
+            if (entry != null)
```

Comment:
is the null check really required here? can't you just use the return value of TryGetValue which tells you if the items exists in the data structure?